### PR TITLE
[MNOE-170] - Adding apps from the SPA

### PR DESCRIPTION
**Pull request by clemthenem**
Added some improvements to add an app from the marketplace directly from the SPA.
I covered the following cases:
- The user has only organisations where he is not allowed to purchase apps.
- The user has multiple organisations but is not allowed to purchase an app on the current organisation.
- The user is allowed to add an app to the current organisation.

PS: Make sure that you merge the pull request that I created for mno-enterprise
-> Adds an app instance endpoint.

### CODE REVIEW

**Comment by clemthenem** [comment]
Added a modal for apps that need customization to connect.
Called from the marketplace and from the dashboard app.

File: src/app/views/apps/modals/app-connect-modal-myob.html [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<div id="app-connect-modal">
+  <div class="modal-header">
+    <div class="close" type="button" aria-label="Close" ng-click="close()">×</div>
+    <h3 class="modal-title">Connect to {{app.name}}</h3>
+  </div>
+  <div class="modal-body">
+    <h4 class="align-center">{{ "mno_enterprise.webhook.o_auth.providers.myob.which_version_used" | translate }}</h4>
+    <div class="spacer1"></div>
```
Comment (alexnoox) [inline src/app/views/apps/modals/app-connect-modal-myob.html]: We do not use `spacer*` classes anymore. Use `.top-buffer-*`. Please remove them everywhere.

File: src/app/views/apps/modals/app-connect-modal-xero.html [hunk 2]
```diff
@@ -0,0 +1,35 @@
+<div id="app-connect-modal">
+  <div class="modal-header">
+    <div class="close" type="button" aria-label="Close" ng-click="close()">×</div>
+    <h3 class="modal-title">Connect to {{app.name}}</h3>
```
Comment (alexnoox) [inline src/app/views/apps/modals/app-connect-modal-xero.html]: i18n

File: src/app/views/apps/modals/app-connect-modal.coffee [hunk 3]
```diff
@@ -0,0 +1,30 @@
+angular.module 'mnoEnterpriseAngular'
+.controller('DashboardAppConnectModalCtrl', ($scope, $uibModalInstance, app, $httpParamSerializer, $window, $http) ->
```
Comment (alexnoox) [inline src/app/views/apps/modals/app-connect-modal.coffee]: Please order this way
```
.controller('DashboardAppConnectModalCtrl', ($scope, $window, $http, $httpParamSerializer, $uibModalInstance, app)
```

File: src/app/views/apps/modals/app-connect-modal.less [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#app-connect-modal {
+  .align-center {
+    text-align: center;
+  }
+
+  .spacer1 {
```
Comment (alexnoox) [inline src/app/views/apps/modals/app-connect-modal.less]: To remove

File: src/app/views/apps/modals/app-connect-modal.less [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#app-connect-modal {
+  .align-center {
```
Comment (alexnoox) [inline src/app/views/apps/modals/app-connect-modal.less]: To remove.

File: src/app/views/apps/modals/app-connect-modal-xero.html [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  </div>
+
+  <div class="modal-body">
+    <h4 class="align-center">{{ "mno_enterprise.webhook.o_auth.providers.xero.how_do_you_want_to_connect" | translate }}</h4>
```
Comment (alexnoox) [inline src/app/views/apps/modals/app-connect-modal-xero.html]: You can use bootstrap class `.text-center`

File: src/app/views/marketplace/marketplace-app.controller.coffee [hunk 7]
```diff
@@ -43,20 +86,48 @@ angular.module 'mnoEnterpriseAngular'
...
+          vm.openChooseAppModal()
+
+      vm.launchAppInstance = ->
         $window.open("/mnoe/launch/#{vm.appInstance.uid}", '_blank')
 
+      #====================================
+      # App Connect modal
+      #====================================
+      vm.connectAppInstance = ->
+        templateUrl = switch
```
Comment (alexnoox) [inline src/app/views/marketplace/marketplace-app.controller.coffee]: Cf previous comment

File: src/app/views/marketplace/marketplace-app.controller.coffee [hunk 8]
```diff
@@ -43,20 +86,48 @@ angular.module 'mnoEnterpriseAngular'
...
+              "INSTALLED_CONNECT"
+            else
+              "INSTALLED_LAUNCH"
         else
           if vm.conflictingAppInstance
             "CONFLICT"
           else
             "INSTALLABLE"
 
+
```
Comment (alexnoox) [inline src/app/views/marketplace/marketplace-app.controller.coffee]: Remove this line.

File: src/app/views/marketplace/marketplace-app.controller.coffee [hunk 9]
```diff
@@ -43,20 +86,48 @@ angular.module 'mnoEnterpriseAngular'
           if vm.app.multi_instantiable
             "INSTALLABLE"
           else
-            "INSTALLED"
+            if vm.app.app_nid != 'office-365' && vm.app.stack == 'connector' && !vm.app.oauth_keys_valid
+              "INSTALLED_CONNECT"
+            else
+              "INSTALLED_LAUNCH"
```
Comment (alexnoox) [inline src/app/views/marketplace/marketplace-app.controller.coffee]: `INSTALLED` won't be use anymore. Please update the function comment doc.

File: src/app/views/marketplace/marketplace-app.controller.coffee [hunk 10]
```diff
@@ -75,6 +146,29 @@ angular.module 'mnoEnterpriseAngular'
         cart.isOpen = true
 
       #====================================
+      # Choose App Modal
+      #====================================
+
+      vm.openChooseAppModal = ->
```
Comment (alexnoox) [inline src/app/views/marketplace/marketplace-app.controller.coffee]: Should be `vm.openChooseOrgoModal`?

File: src/app/views/marketplace/marketplace-app.controller.coffee [hunk 11]
```diff
@@ -75,6 +146,29 @@ angular.module 'mnoEnterpriseAngular'
...
+        vm.chooseAppModalInstance = $uibModal.open({
+          backdrop: 'static'
+          size: 'lg'
+          templateUrl: 'app/views/marketplace/modals/choose-orga-modal.html'
+          windowClass: 'inverse'
+          scope: $scope
```
Comment (alexnoox) [inline src/app/views/marketplace/marketplace-app.controller.coffee]: Can you put the modal related functions in its own controller?

File: src/app/views/marketplace/modals/choose-orga-modal.html [hunk 12]
```diff
@@ -0,0 +1,60 @@
+
+<!-- Modal Header -->
+<div class="modal-header">
+  <div class="close" type="button" ng-click="vm.cancelChooseAppModal()" >×</div>
+</div>
+<!-- Modal Body -->
+<div class="modal-body">
+
+  <div class="row" ng-show="!modal.isLoading()" >
```
Comment (alexnoox) [inline src/app/views/marketplace/modals/choose-orga-modal.html]: Where is `modal.isLoading()` defined?

File: src/app/components/dashboard-apps-dock/dashboard-apps-dock.coffee [hunk 13]
```diff
@@ -76,6 +76,24 @@ DashboardAppsDockCtrl = ($scope, $cookies, $uibModal, $window, MnoeOrganizations
     )
 
   #====================================
+  # App Connect modal
+  #====================================
+  $scope.showConnectModal = (app) ->
+    templateUrl = switch
+      when app.app_nid == "xero" then "app/views/apps/modals/app-connect-modal-xero.html"
```
Comment (clemthenem) [inline src/app/components/dashboard-apps-dock/dashboard-apps-dock.coffee]: Cool! Thanks for the tip :)

**Comment by clemthenem** [comment]
I added the changes to the pull request! 
Let me know if you need me to change anything.

File: src/app/components/dashboard-apps-dock/dashboard-apps-dock.html [hunk 14]
```diff
@@ -13,7 +13,7 @@
               <a ng-click="launchAction(app,$event);"><i class="fa fa-play-circle"></i>{{ 'mno_enterprise.templates.impac.dock.launch' | translate }}</a>
             </div>
             <div class="menu-item ng-hide" ng-show="helper.isOauthConnectBtnShown(app)">
-                <a href="" ng-click="helper.oAuthConnectPath(app)"><i class="fa fa-play-circle" >
+                <a ng-click="showConnectModal(app)"><i class="fa fa-play-circle" >
```
Comment (clemthenem) [inline src/app/components/dashboard-apps-dock/dashboard-apps-dock.html]: It doesn't :)

**Comment by clemthenem** [comment]
Rebased from master! :)

**Pull request merged**